### DISCUSSION TITLE
Support for Orange Pi Zero 3

### DISCUSSION THREAD

**Discussion by tusc** [marcone/teslausb#785] [General]
Has anyone tried this SBC lately? I just purchased it off Amazon for a reasonable price ($28 USD for a 2GB model). I initially tried Teslausb on a Radxa Zero but WiFi was horrendously slow. The Orange PI has better WiFi performance and offers similar power consumption, about .9 watts at idle.

I installed the Ubuntu server image from Orange with the 6.1.31 kernel:
```
orangepi@teslausb:~$ uname -a
Linux teslausb 6.1.31-sun50iw9 #1.0.0 SMP Mon Jul  3 13:52:39 CST 2023 aarch64 aarch64 aarch64 GNU/Linux
orangepi@teslausb:~$ cat /etc/issue
Orange Pi 1.0.0 Jammy \l
```

I went ahead and ran the Teslausb install.sh and finished with rc.local. The 64GB sdcard was resized and installation finished without issue. It appears that OTG is supported. I have the device attached to my Macbook via one of the usb-c ports and can see the TeslaCam and TeslaTrackMode subdirectories on the CAM external drive.

Should this device be qualified given its efficient power usage and decent wifi speeds?

**Comment by marcone** [comment]
>  I initially tried Teslausb on a Radxa Zero but WiFi was horrendously slow

What speed are we talking about here, and did you get one of the Radxa Zero variants that supports 5 GHz wifi?
Did you have an antenna attached to it?

> Should this device be qualified given its efficient power usage and decent wifi speeds?

Not sure what you mean by "qualified" here.

**Comment by tusc** [comment]
> > I initially tried Teslausb on a Radxa Zero but WiFi was horrendously slow
> 
> What speed are we talking about here, and did you get one of the Radxa Zero variants that supports 5 GHz wifi? Did you have an antenna attached to it?
> 
It's the Radxa Zero w/ 2GB, WiFI 5ghz with support for an external antenna. I ordered it from allnet along with the regular and short PCB WiFi antenna. Right now it's being tested with the regular length antenna. Using the latest build for Armbian. The best I can get for network speed test from teslausb is 39Mb/sec. I even attempted to attach a USB Gigabit ethernet card from a USB-C dongle and could barely hit 300Mb/sec from that. Not sure what could be slowing it down.

here's the output from iwconfig:
```
root@teslausb:~# iwconfig
lo        no wireless extensions.

wlan0     IEEE 802.11  ESSID:"XXX"
          Mode:Managed  Frequency:5.5 GHz  Access Point: XX:XX:XX:XX:XX:XX
          Bit Rate=325 Mb/s   Tx-Power=31 dBm
          Retry short limit:7   RTS thr:off   Fragment thr:off
          Encryption key:off
          Power Management:off
          Link Quality=70/70  Signal level=-38 dBm
          Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
          Tx excessive retries:0  Invalid misc:0   Missed beacon:0

```

iperf3 also shows simliar speed test results as the teslsausb network test tool:

```
root@teslausb:~# iperf3 -s
-----------------------------------------------------------
Server listening on 5201 (test #1)
-----------------------------------------------------------
Accepted connection from 192.168.1.33, port 51531
[  5] local 192.168.1.135 port 5201 connected to 192.168.1.33 port 51532
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  4.58 MBytes  38.4 Mbits/sec
[  5]   1.00-2.00   sec  4.97 MBytes  41.7 Mbits/sec
[  5]   2.00-3.00   sec  4.86 MBytes  40.8 Mbits/sec
[  5]   3.00-4.00   sec  4.95 MBytes  41.5 Mbits/sec
[  5]   4.00-5.00   sec  4.72 MBytes  39.6 Mbits/sec
[  5]   5.00-6.00   sec  4.97 MBytes  41.7 Mbits/sec
[  5]   6.00-7.00   sec  4.83 MBytes  40.5 Mbits/sec
[  5]   7.00-8.00   sec  5.12 MBytes  43.0 Mbits/sec
[  5]   8.00-9.00   sec  4.76 MBytes  39.9 Mbits/sec
[  5]   9.00-10.00  sec  4.67 MBytes  39.2 Mbits/sec
[  5]  10.00-10.03  sec   205 KBytes  61.4 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.03  sec  48.6 MBytes  40.7 Mbits/sec                  receiver
```
> > Should this device be qualified given its efficient power usage and decent wifi speeds?
> 
> Not sure what you mean by "qualified" here.

I should have said, is the OP Zero 3 recommended? Are others using it? I didn't see it listed on the hardware page. Given the affordable price of the OP Zero 3, I'm just surprised no one has brought it up.

**Reply by marcone** [reply]
> The best I can get for network speed test from teslausb is 39Mb/sec. I even attempted to attach a USB Gigabit ethernet card from a USB-C dongle and could barely hit 300Mb/sec from that.

39 Mb/sec isn't terrible I think. I get 56 Mb/s on the 5 GHz band under optimal conditions, and much, much worse on 2.4 GHz (as low as 5 Mbit/s)
For ethernet my guess would be you ran into CPU and/or USB bottlenecks.

> is the OP Zero 3 recommended? Are others using it? I didn't see it listed on the hardware page. Given the affordable price of the OP Zero 3, I'm just surprised no one has brought it up

Perhaps because it's relatively new (released just 3 months ago).

**Comment by tusc** [comment]
> > The best I can get for network speed test from teslausb is 39Mb/sec. I even attempted to attach a USB Gigabit ethernet card from a USB-C dongle and could barely hit 300Mb/sec from that.
> 
> 39 Mb/sec isn't terrible I think. I get 56 Mb/s on the 5 GHz band under optimal conditions, and much, much worse on 2.4 GHz (as low as 5 Mbit/s) For ethernet my guess would be you ran into CPU and/or USB bottlenecks.

Here's the wifi test for the OP 3: 
![image](https://github.com/marcone/teslausb/assets/9438017/c676c30b-6e2f-41ad-b44f-dbaae8a9c774)

When testing the ethernet port I can max out at 900Mb/sec.

> 
> > is the OP Zero 3 recommended? Are others using it? I didn't see it listed on the hardware page. Given the affordable price of the OP Zero 3, I'm just surprised no one has brought it up
> 
> Perhaps because it's relatively new (released just 3 months ago).

Understood.

**Reply by marcone** [reply]
The Orange Pi Zero 3 seems to underperform in terms of USB drive emulation:

![image](https://github.com/marcone/teslausb/assets/48169102/1cfe26e0-0c1b-4270-a08f-9e28962e9419)

The sequential write numbers are the important ones here, and the Orange Pi Zero 3 is quite a bit slower than Raspberry Pi Zero (see https://github.com/marcone/teslausb/issues/239#issuecomment-979417606 for Raspberry Pi numbers). Not sure if that matters in practice though.

**Reply by tusc** [reply]
I saw similar results from my Macbook connected to the OPI Zero 3 via USB-C. According to the Tesla manual, they require at least 4MB/sec throughput. I think the advantage of this SBC is the increased WiFi throughput (IMHO).

![image](https://github.com/marcone/teslausb/assets/9438017/c14f210d-1c88-4e09-b3d6-e933967f346a)

**Comment by marcone** [comment]
> I installed the Ubuntu server image from Orange with the 6.1.31 kernel:

I tried an Orange Pi Zero 3 1 GB, but the ubuntu image linked from [here](http://www.orangepi.org/html/hardWare/computerAndMicrocontrollers/service-and-support/Orange-Pi-Zero-3.html) doesn't seem to be working. Not sure if that's because the device is broken or because the image for the 1GB version is broken.
I attached serial debug and get this on power-on:

```
U-Boot SPL 2021.07-orangepi (Jul 03 2023 - 13:51:47 +0800)
DRAM: 1024 MiB
Trying to boot from MMC1
NOTICE:  BL31: v2.9(debug):463655cc
NOTICE:  BL31: Built : 16:14:59, Jun  5 2023
NOTICE:  BL31: Detected Allwinner H616 SoC (1823)
NOTICE:  BL31: Found U-Boot DTB at 0x4a0827f8, model: OrangePi Zero3
INFO:    ARM GICv2 driver initialized
INFO:    Configuring SPC Controller
INFO:    PMIC: Probing AXP305 on RSB
ERROR:   RSB: set run-time address: 0x10003
INFO:    Could not init RSB: -65539
INFO:    BL31: Platform setup done
INFO:    BL31: Initializing runtime services
INFO:    BL31: cortex_a53: CPU workaround for 855873 was applied
INFO:    BL31: cortex_a53: CPU workaround for 1530924 was applied
INFO:    PSCI: Suspend is unavailable
INFO:    BL31: Preparing for EL3 exit to normal world
INFO:    Entry point address = 0x4a000000
INFO:    SPSR = 0x3c9
INFO:    Changed devicetree.


U-Boot 2021.07-orangepi (Jul 03 2023 - 13:51:47 +0800) Allwinner Technology

CPU:   Allwinner H616 (SUN50I)
Model: OrangePi Zero3
I2C:   ready
DRAM:  1 GiB
MMC:   mmc@4020000: 0
Loading Environment from FAT... Unable to use mmc 0:1... In:    serial@5000000
Out:   serial@5000000
Err:   serial@5000000
Net:   No ethernet found.
starting USB...
No working controllers found
Autoboot in 1 seconds, press <Space> to stop
switch to partitions #0, OK
mmc0 is current device
Scanning mmc 0:1...
Found U-Boot script /boot/boot.scr
3636 bytes read in 4 ms (887.7 KiB/s)
## Executing script at 4fc00000
Wrong image format for "source" command
SCRIPT FAILED: continuing...
libfdt fdt_check_header(): FDT_ERR_BADMAGIC
Scanning disk mmc@4020000.blk...
Found 2 disks
No EFI system partition
BootOrder not defined
EFI boot manager: Cannot load any image
starting USB...
No working controllers found
USB is stopped. Please issue 'usb start' first.
starting USB...
No working controllers found
No ethernet found.
missing environment variable: pxeuuid
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/00000000
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/0000000
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/000000
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/00000
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/0000
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/000
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/00
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/0
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/default-arm-sunxi-sunxi
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/default-arm-sunxi
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/default-arm
No ethernet found.
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/default
No ethernet found.
Config file not found
starting USB...
No working controllers found
No ethernet found.
No ethernet found.
=>
```

**Reply by marcone** [reply]
The debian server image appears to work though.

**Comment by marcone** [comment]
> I tried an Orange Pi Zero 3 1 GB, but the ubuntu image linked from [here](http://www.orangepi.org/html/hardWare/computerAndMicrocontrollers/service-and-support/Orange-Pi-Zero-3.html) doesn't seem to be working. Not sure if that's because the device is broken or because the image for the 1GB version is broken.

Apparently this was because the Windows app I was using to unzip the archive produced a corrupt file. After switching to the command line `7z`, the resulting .img was OK and booted fine.

**Comment by tusc** [comment]
Are you seeing similar performance numbers to what I saw?

**Reply by marcone** [reply]
For networking? Not really. It seems to prefer connecting to 2.4 GHz wifi, and I'm seeing about 25 Mbit/s with that. When I force it onto 5 GHz wifi I'm seeing up to120 MBit/s if I hold the antenna _just_ right, but with the antenna just lying flat on the desk next to the board it's around 45 Mbit/s. Also seeing quite a bit of packet loss in that case.

**Comment by tusc** [comment]
Fyi,

now that Armbian has preliminary support for the Orange PI Zero 3, I went ahead and tried to load the latest version of Teslausb. Fortunately it ran without issue. This is the current wifi speed test:

<img width="422" alt="image" src="https://github.com/marcone/teslausb/assets/9438017/fa078a49-906b-4a5e-9dfc-d80aeb35353c">

Armbian download images at the bottom of this page: https://www.armbian.com/orange-pi-zero-3/

**Comment by marcone** [comment]
> now that Armbian has preliminary support for the Orange PI Zero 3

Where do you see this? I only see that it is "community supported" on https://www.armbian.com/orange-pi-zero-3/ and there are no installation images linked from that page as far as I can see.

**Comment by tusc** [comment]
There used to be rolling release images at that link as pointed out by the post below. Not sure why they were taken down. I downloaded one last week.

https://forum.armbian.com/topic/29202-orange-pi-zero-3/?do=findComment&comment=180779

**Comment by tusc** [comment]
Looks like the CI job failed to update the web site but you can grab an image from here:  https://github.com/armbian/os/releases/download/24.2.0-trunk.519/Armbian_24.2.0-trunk.519_Orangepizero3_trixie_current_6.6.15_minimal.img.xz

Mentioned here:  https://forum.armbian.com/topic/29202-orange-pi-zero-3/?do=findComment&comment=180921

**Comment by tusc** [comment]
Download links are back on the page: https://www.armbian.com/orange-pi-zero-3/

**Reply by gsj585** [reply]
你好  使用Orange Pi Zero 3在中国怎么teslausbPublic使用阿里云同步，有详细的教程吗  可以教教我吗      

Can you teach me

**Comment by gsj585** [comment]
你好Orange Pi Zero 3安装怎么teslausbPublic详细教程有吗  可以配置使用阿里云同步吗

**Comment by gsj585** [comment]
Armbian_24.2.0-trunk.519_Orangepizero3_trixie_current_6.6.15_minimal.img.xz


这个版本下载不了   可以使用Armbian_25.5.1_Orangepizero3_noble_current_6.12.23.img版本吗？有大神测试吗

**Comment by gsj585** [comment]
Using the Orangepizero3 device with a 16G memory card and a 128G USB drive, when there is a new video in the virtual disk, executing the "make_snapshot.sh" script fails to obtain the new snapshot. Only by restarting the device can the new snapshot be obtained. How to solve this problem?

**Reply by marcone** [reply]
What exactly do you mean by "fails to obtain the new snapshot"?

**Reply by gsj585** [reply]
yes   


  > What exactly do you mean by "fails to obtain the new snapshot"?

**Reply by gsj585** [reply]
Using the 16G memory card and 128G USB drive solution, when there are new videos in the virtual disk, executing "make_snapshot.sh" will not be able to obtain the new snapshots. Only after disconnecting the virtual disk from the vehicle console can the archiving synchronization be triggered, and it will be effective.

Using an external USB drive, I can't transfer the latest snapshot.


> What exactly do you mean by "fails to obtain the new snapshot"?

**Reply by marcone** [reply]
Again: what exactly do you mean by "fails to obtain the new snapshot"? How does it fail to take the snapshot? What goes wrong? Did you look at the logs? What do they say?
